modules/sockets/sockets.cpp

```diff
@@ -20,228 +20,299 @@
-    fd_set rfds;
-    struct timeval tv;
-    // Fill in ...
-    FD_ZERO(&rfds);
-    FD_SET(socket, &rfds);
-    tv.tv_sec = 0;
-    tv.tv_usec = timeout;
-    // Now we "select", which will show us if new data is waiting in the socket's buffer
-    if (select(socket+1, &rfds, NULL, NULL, &tv) > 0) 
-       return 1; // Ok, new data, return it
-     else 
-       return 0; // No new data, return it
-}
-
-// native socket_recv(_socket, _data[], _length);
-static cell AMX_NATIVE_CALL socket_recv(AMX *amx, cell *params)  /* 2 param */
+	SOCKET_TCP = 1,
+	SOCKET_UDP
+};
+
+// socket_open(Hostname[], Port, Protocol = SOCKET_TCP, &errorNum)
+//
+static cell AMX_NATIVE_CALL socket_open(AMX * pAmx, cell * pParams)
 {
-    int socket = params[1];
-    int length = params[3];
-    int tmp = -1;
-    // First we dynamicly allocate a block of heap memory for recieving our data
-    char *tmpchar = new char[length];
-    if(tmpchar == NULL) return -1; // If we didn't got a block, we have to quit here to avoid sigsegv
-    // And set it all to 0, because the memory could contain old trash
-    memset(tmpchar, 0, length);
-    // Now we recieve
-    tmp = recv(socket, tmpchar, length-1, 0);
-	if (tmp == -1)
+	// Gets hostname, hostname's length and error address.
+	//
+	int Length = 0;
+	char * pName = MF_GetAmxString(pAmx, pParams[1], 0, &Length);
+	cell * pError = MF_GetAmxAddr(pAmx, pParams[4]);
+
+	// Hostname parameter is empty.
+	//
+	if (Length < 1)
 	{
-		delete [] tmpchar;
+		*pError = 2;
 		return -1;
-	}
-    // And put a copy of our recieved data into amx's string
-    tmpchar[tmp]='\0';
-    int nlen = 0;
-    //int max = params[3];
-    int max = length-1;
-    const char* src = tmpchar;
-    cell* dest = MF_GetAmxAddr(amx,params[2]);
-    while(max--&&nlen<tmp){
-     *dest++ = (cell)*src++;
-     nlen++;
-    }
-    *dest = 0;
-    // And we need to free up the space to avoid wasting memory
-    delete [] tmpchar;
-    // And finnally, return the what recv returnd
-    return tmp;
-}
-
-// native socket_send(_socket, _data[], _length);
-static cell AMX_NATIVE_CALL socket_send(AMX *amx, cell *params)  /* 3 param */
+	};
+
+	// Opens new socket.
+	//
+	int Socket = (int)socket(AF_INET, AmxxSocketType(pParams[3]) == SOCKET_TCP ? SOCK_STREAM : SOCK_DGRAM, 0);
+	if (Socket < 0)
+	{
+		*pError = 1;
+		return -1;
+	};
+
+	// Server information.
+	//
+	struct sockaddr_in Server;
+	memset(&Server, 0, sizeof Server);
+
+	// Gets address.
+	//
+	unsigned long Addr = 0UL;
+	if ((Addr = inet_addr(pName)) != INADDR_NONE)
+		memcpy((char *)&Server.sin_addr, &Addr, sizeof Addr);
+
+	else
+	{
+		struct hostent * pInfo = gethostbyname(pName);
+		if (!pInfo)
+		{
+			*pError = 2;
+			return -1;
+		};
+
+		memcpy((char *)&Server.sin_addr, pInfo->h_addr, pInfo->h_length);
+	};
+
+	Server.sin_family = AF_INET;
+	Server.sin_port = htons((unsigned short)pParams[2]);
+
+	// Connects.
+	//
+	if (connect(Socket, (struct sockaddr *)&Server, sizeof Server) < 0)
```

Comment:
If the connection failed after the socket has been created, just close the socket and return -1.
The socket must not stay initialized.